ignite/contrib/handlers/lr_finder.py

```diff
@@ -0,0 +1,425 @@
+            self._engine = None
+            if engine.has_event_handler(self._run, Events.STARTED):
+                engine.remove_event_handler(self._run, Events.STARTED)
+            if engine.has_event_handler(self._warning, Events.COMPLETED):
+                engine.remove_event_handler(self._warning, Events.COMPLETED)
+            if engine.has_event_handler(self._reset, Events.COMPLETED):
+                engine.remove_event_handler(self._reset, Events.COMPLETED)
+        else:
+            warnings.warn("This LRFinder isn't attached, this action has no effect")
+
+    def get_results(self):
+        """
+        Returns: dictionary with loss and lr logs fromm the previous run
+        """
+        return self._history
+
+    def plot(self, skip_start=10, skip_end=5, log_lr=True):
+        """Plots the learning rate range test.
+
+        Args:
+            skip_start (int, optional): number of batches to trim from the start.
+                Default: 10.
+            skip_end (int, optional): number of batches to trim from the start.
+                Default: 5.
+            log_lr (bool, optional): True to plot the learning rate in a logarithmic
+                scale; otherwise, plotted in a linear scale. Default: True.
+        """
+
+        if skip_start < 0:
+            raise ValueError("skip_start cannot be negative")
+        if skip_end < 0:
+            raise ValueError("skip_end cannot be negative")
+
+        # Get the data to plot from the history dictionary. Also, handle skip_end=0
+        # properly so the behaviour is the expected
+        try:
+            import matplotlib.pyplot as plt
+
+            lrs = self._history["lr"]
+            losses = self._history["loss"]
+            if skip_end == 0:
+                lrs = lrs[skip_start:]
+                losses = losses[skip_start:]
+            else:
+                lrs = lrs[skip_start:-skip_end]
+                losses = losses[skip_start:-skip_end]
+
+            # Plot loss as a function of the learning rate
+            plt.plot(lrs, losses)
+            if log_lr:
+                plt.xscale("log")
+            plt.xlabel("Learning rate")
+            plt.ylabel("Loss")
+            plt.show()
+
+        except ModuleNotFoundError:
+            self._logger.warning("matplotlib not found, can't plot result")
+
+    def lr_suggestion(self):
+        """
+        Returns: learning rate at the minimum numerical gradient
+        """
+        loss = self._history["loss"]
+        grads = [loss[i] - loss[i - 1] for i in range(1, len(loss))]
+        min_grad_idx = np.argmin(grads) + 1
+        return self._history["lr"][int(min_grad_idx)]
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, val, tb):
+        self.detach()
+
+
+class AlreadyAttachedError(Exception):
+    """
+    Exception class to raise if trying to attach an already attached LRFinder.
+    """
+
+
+class NotEnoughIterationWarning(Warning):
```

Comment:
Previously, we used `UserWarning` as warning types, I think can go without custom warnings.

Reply:
No Problem

Reply:
Done